src/proxyConnect.ts

```diff
@@ -0,0 +1,105 @@
+    /** milliseconds to wait before giving up */
+    private _timeout: number
+    public msgs: Map<string, string>
+    private _isRegistered = false
+    private _resolve: Function
+
+    constructor(host = '127.0.0.1', port = 9001, allowMultipleSessions = 1, key = DEFAULTIDEKEY, timeout = 3000) {
+        super()
+        this._allowMultipleSessions = allowMultipleSessions
+        this._host = host
+        this._key = key
+        this._port = port
+        this._timeout = timeout
+        this.msgs = new Map([
+            ['deregisterInfo', `Deregistering ${this._key} with proxy @ ${this._host}:${this._port}`],
+            ['deregisterSuccess', 'Deregistration successful'],
+            ['duplicateKey', 'IDE Key already exists'],
+            ['nonexistentKey', 'No IDE key'],
+            ['registerInfo', `Registering ${this._key} with proxy @ ${this._host}:${this._port}`],
+            ['registerSuccess', 'Registration successful'],
+            ['resolve', `Failure to resolve ${this._host}`],
+            ['timeout', `Timeout connecting to ${this._host}:${this._port}`],
+            ['defaultError', 'Unknown Error'],
+        ])
+        this._socket.on('error', (err: Error) => {
+            // Propagate error up
+            this._socket.end()
+            this.emit('error', err instanceof Error ? err : new Error(err))
+        })
+        this._socket.on('lookup', (err: Error | null, address: string, family: string | null, host: string) => {
+            if (err instanceof Error) {
+                this._socket.emit('error', `${err.message}${address || host || ''}`)
+            }
+        })
+        this._socket.on('data', data => this._responseStrategy(data))
+        this._socket.setTimeout(this._timeout)
+        this._socket.on('timeout', () => {
+            this._socket.emit('error', this.msgs.get('timeout'))
+        })
+    }
+
+    private _command(cmd: string, msg?: string) {
+        this.emit('info', msg)
+        this._socket.connect(this._port, this._host, () => this._socket.end(cmd))
+    }
+
+    /** Register/Couples ideKey to IP so the proxy knows who to send what */
+    public sendProxyInitCommand() {
+        if (!this._isRegistered) {
+            this._command(
+                `proxyinit -k ${this._key} -p ${this._port} -m ${this._allowMultipleSessions}`,
+                this.msgs.get('registerInfo')
+            )
+        }
+    }
+
+    /** Deregisters/Decouples ideKey from IP, allowing others to use the ideKey */
+    public sendProxyStopCommand(resolve: Function) {
+        if (this._isRegistered) {
+            this._resolve = resolve
+            this._command(`proxystop -k ${this._key}`, this.msgs.get('deregisterInfo'))
+        }
+    }
+
+    /** Parse data from response server and emit the relevant notification. */
+    private _responseStrategy(data: Buffer) {
+        const documentElement = this._parser.parseFromString(decode(data, ENCODING), 'application/xml').documentElement
```

Comment:
So the XDGP Proxy seems to speak XDGP like the XDebug connection itself, i.e. accept CLI commands and return XML? Why does this have it's completely parsing logic from raw Buffers, Sockets and XML instead of sharing code with `DBGPConnection` or even `XDebugConnection`? For example, in XDebugConnection, every possible packet is represented as a class.

Reply:
They do parallel in functionality. Thus we initially thought of using the `DBGPConnection` and subsequently the `XDebugConnection` classes but realized the value of reusing their code was of small value compared to the incurred overhead.

The xdebug documentation states that proxy connection is to be terminated upon ide key registration; therefore they must be separate connections. In addition, the returned xml is structured differently. As we continued to modify the existing data handler and connection lifecycle, it became unnecessarily complicated and the refactoring seemed more like code smells than not.

Reply:
What is the overhead/complexity? Do you maybe have an older commit I can look at that?

Reply:
The overhead primarily lay in that the `DbgpConnection`, and its subclass `Connection`, are written for a connection whose intent is to continuously communicate with the server. With this comes many aspects: various connection states, commands queueing, and a range of potential content; none of which `ProxyConnection` partakes in. It has **3** jobs: register / deregister a users ide(key) with the proxy server and tell you if it's succeeded. In addition, the former part occurs with *2* separate instances of the connection.

`DbgpConnection` and `Connection` are approximately 120% and %375, respectively, larger than `ProxyConnection`. Subclassing either would have simply bloated the extension.

---

In terms of commits, I took at our local branches and it doesn't look like we have those commits readily available. Digging through the reflog could result in *something* but most likely not.